base_layer/core/src/chain_storage/blockchain_database.rs

```diff
@@ -1471,20 +1475,55 @@ fn rewind_to_height<T: BlockchainBackend>(db: &mut T, height: u64) -> Result<Vec
         last_block_height,
         last_block_height - steps_back
     );
+
+    let mut prune_past_horizon = false;
+    if steps_back > metadata.pruning_horizon() && metadata.pruning_horizon() > 0 {
+        info!(target: LOG_TARGET, "WARNING, reorg past pruning horizon");
+        info!(target: LOG_TARGET, "WARNING, rewinding back to 0");
+        steps_back = metadata.pruning_horizon();
+        prune_past_horizon = true;
+        height = 0;
+    }
+    let (last_header, header_accumulated_data) = db.fetch_header_and_accumulated_data(height)?;
+
     for h in 0..steps_back {
+        debug!(target: LOG_TARGET, "Deleting block {}", last_block_height - h,);
         let block = fetch_block(db, last_block_height - h)?;
-        if block.is_pruned() {
-            unimplemented!("Rewinding past the pruning horizon not implemented.");
-        }
         let block = Arc::new(block.clone().try_into_chain_block()?);
         txn.delete_block(block.block.hash());
         txn.delete_header(last_block_height - h);
-        txn.insert_chained_orphan(block.clone());
+        if !prune_past_horizon {
+            // Because we know we will remove blocks we cant recover, this will be a destructive rewind, so we cant
+            // recover from this apart from resync from another peer. Failure here should not be common as
+            // this chain has a valid proof of work that has been tested at this point in time.
+            txn.insert_chained_orphan(block.clone());
+        }
         removed_blocks.push(block);
     }
 
+    if prune_past_horizon {
+        // We are rewinding past pruning horizon, so we need to remove all blocks and the UTXO's from them. We do not
+        // have to delete the headers as they are still valid.
+        for h in 0..(last_block_height - steps_back) {
+            debug!(
+                target: LOG_TARGET,
+                "Deleting blocks and utxos {}",
+                last_block_height - h - steps_back,
+            );
+            let block = fetch_block(db, last_block_height - h - steps_back)?;
+            let block = Arc::new(block.clone().into_pruned_chain_block());
+            txn.delete_block(block.block.hash());
+            removed_blocks.push(block);
```

Comment:
This is potentially a LOT of blocks in memory

Reply:
Yip, I am in two minds on this, keep it like this, or just don't give these to the mempool. Or we trigger a mempool wipe or the like.